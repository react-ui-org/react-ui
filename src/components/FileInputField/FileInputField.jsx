import PropTypes from 'prop-types';
import React, {
  useCallback,
  useEffect,
  useContext,
  useImperativeHandle,
  useRef,
  useState,
} from 'react';
import { withGlobalProps } from '../../providers/globalProps';
import { classNames } from '../../helpers/classNames';
import { transferProps } from '../../helpers/transferProps';
import { TranslationsContext } from '../../providers/translations';
import { getRootSizeClassName } from '../_helpers/getRootSizeClassName';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { resolveContextOrProp } from '../_helpers/resolveContextOrProp';
import { InputGroupContext } from '../InputGroup';
import { Text } from '../Text';
import { FormLayoutContext } from '../FormLayout';
import styles from './FileInputField.module.scss';

/* istanbul ignore file */

export const FileInputField = React.forwardRef((props, ref) => {
  const {
    disabled,
    fullWidth,
    helpText,
    id,
    isLabelVisible,
    label,
    layout,
    multiple,
    onFilesChanged,
    required,
    size,
    validationState,
    validationText,
    ...restProps
  } = props;

  const formLayoutContext = useContext(FormLayoutContext);
  const inputGroupContext = useContext(InputGroupContext);
  const translations = useContext(TranslationsContext);

  const [selectedFileNames, setSelectedFileNames] = useState([]);
  const [isDragging, setIsDragging] = useState(false);

  const internalInputRef = useRef();

  const handleReset = useCallback((event) => {
    if (internalInputRef.current) {
      internalInputRef.current.value = '';
    }

    setSelectedFileNames([]);
    onFilesChanged([], event);
  }, [onFilesChanged]);

  // We need to have a reference to the input element to be able to call its methods,
  // but at the same time we want to expose this reference to the parent component in
  // case someone wants to call input methods from outside the component.
  useImperativeHandle(
    ref,
    () => {
      // The reason of extending object instead of using spread operator is that
      // if it is transformed to the object, it changes the reference of the object
      // and its prototype chain.
      const inputEl = internalInputRef?.current ?? {};
      inputEl.resetState = () => {
        handleReset(null);
      };
      return inputEl;
    },
    [handleReset],
  );

  const handleFileChange = (files, event) => {
    if (files.length === 0) {
      setSelectedFileNames([]);
      onFilesChanged([], event);
      return;
    }

    // Mimic the native behavior of the `input` element: if multiple files are selected and the input
    // does not accept multiple files, no files are processed.
    if (files.length > 1 && !multiple) {
      setSelectedFileNames([]);
      onFilesChanged([], event);
      return;
    }

    const fileNames = [];

    [...files].forEach((file) => {
      fileNames.push(file.name);
    });

    setSelectedFileNames(fileNames);
    onFilesChanged(files, event);
  };

  const handleInputChange = (event) => {
    handleFileChange(event.target.files, event);
  };

  const handleClick = () => {
    internalInputRef?.current.click();
  };

  const handleDrop = (event) => {
    event.preventDefault();
    handleFileChange(event.dataTransfer.files, event);
    setIsDragging(false);
  };

  const handleDragOver = (event) => {
    if (!isDragging) {
      setIsDragging(true);
    }
    event.preventDefault();
  };

  const handleDragLeave = () => {
    if (isDragging) {
      setIsDragging(false);
    }
  };

  useEffect(() => {
    const inputEl = internalInputRef.current;
    if (!inputEl) {
      return () => {};
    }

    const { form } = inputEl;
    if (!form) {
      return () => {};
    }

    form.addEventListener('reset', handleReset);

    return () => {
      form.removeEventListener('reset', handleReset);
    };
  }, [handleReset]);

  return (
    <div
      className={classNames(
        styles.root,
        fullWidth && styles.isRootFullWidth,
        formLayoutContext && styles.isRootInFormLayout,
        resolveContextOrProp(formLayoutContext && formLayoutContext.layout, layout) === 'horizontal'
          ? styles.isRootLayoutHorizontal
          : styles.isRootLayoutVertical,
        resolveContextOrProp(inputGroupContext && inputGroupContext.disabled, disabled) && styles.isRootDisabled,
        inputGroupContext && styles.isRootGrouped,
        isDragging && styles.isRootDragging,
        required && styles.isRootRequired,
        getRootSizeClassName(
          resolveContextOrProp(inputGroupContext && inputGroupContext.size, size),
          styles,
        ),
        getRootValidationStateClassName(validationState, styles),
      )}
      id={`${id}__root`}
      onDragLeave={!disabled ? handleDragLeave : undefined}
      onDragOver={!disabled ? handleDragOver : undefined}
      onDrop={!disabled ? handleDrop : undefined}
    >
      <label
        className={classNames(
          styles.label,
          (!isLabelVisible || inputGroupContext) && styles.isLabelHidden,
        )}
        htmlFor={id}
        id={`${id}__labelText`}
      >
        {label}
      </label>
      <div className={styles.field}>
        <div className={styles.inputContainer}>
          <input
            {...transferProps(restProps)}
            className={styles.input}
            disabled={resolveContextOrProp(inputGroupContext && inputGroupContext.disabled, disabled)}
            id={id}
            multiple={multiple}
            onChange={handleInputChange}
            ref={internalInputRef}
            required={required}
            tabIndex={-1}
            type="file"
          />
          <button
            className={styles.dropZone}
            disabled={resolveContextOrProp(inputGroupContext && inputGroupContext.disabled, disabled)}
            onClick={handleClick}
            type="button"
          >
            <Text lines={1}>
              {isDragging && (
                <span className={styles.dropFileHereText}>
                  {translations.FileInputField.dropFileHere}
                </span>
              )}
              {!isDragging && !selectedFileNames.length && (
                <>
                  <span className={styles.dropZoneLink}>{translations.FileInputField.browse}</span>
                  {' '}
                  {translations.FileInputField.drop}
                </>
              )}
              {!isDragging && selectedFileNames.length === 1 && selectedFileNames[0]}
              {!isDragging && selectedFileNames.length > 1 && (
                <>
                  {selectedFileNames.length}
                  {' '}
                  {translations.FileInputField.filesSelected}
                </>
              )}
            </Text>
          </button>
        </div>
        {helpText && (
          <div
            className={styles.helpText}
            id={`${id}__helpText`}
          >
            {helpText}
          </div>
        )}
        {validationText && (
          <div
            className={styles.validationText}
            id={`${id}__validationText`}
          >
            {validationText}
          </div>
        )}
      </div>
    </div>
  );
});

FileInputField.defaultProps = {
  disabled: false,
  fullWidth: false,
  helpText: null,
  isLabelVisible: true,
  layout: 'vertical',
  multiple: false,
  required: false,
  size: 'medium',
  validationState: null,
  validationText: null,
};

FileInputField.propTypes = {
  /**
   * If `true`, the input will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * If `true`, the field will span the full width of its parent.
   */
  fullWidth: PropTypes.bool,
  /**
   * Optional help text.
   */
  helpText: PropTypes.node,
  /**
   * ID of the `<input>` HTML element.
   *
   * Also serves as base for ids of nested elements:
   * * `<ID>__label`
   * * `<ID>__labelText`
   * * `<ID>__helpText`
   * * `<ID>__validationText`
   */
  id: PropTypes.string.isRequired,
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible: PropTypes.bool,
  /**
   * File input field label.
   */
  label: PropTypes.node.isRequired,
  /**
   * Layout of the field.
   *
   * Ignored if the component is rendered within `FormLayout` component
   * as the value is inherited in such case.
   *
   */
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  /**
   * If `true`, the input will accept multiple files.
   */
  multiple: PropTypes.bool,
  /**
   * Callback fired when the value of the input changes.
   */
  onFilesChanged: PropTypes.func.isRequired,
  /**
   * If `true`, the input will be required.
   */
  required: PropTypes.bool,
  /**
   * Size of the field.
   *
   * Ignored if the component is rendered within `InputGroup` component as the value is inherited in such case.
   */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  /**
   * Validation message to be displayed.
   */
  validationText: PropTypes.node,
};

export const FileInputFieldWithGlobalProps = withGlobalProps(FileInputField, 'FileInputField');

export default FileInputFieldWithGlobalProps;
