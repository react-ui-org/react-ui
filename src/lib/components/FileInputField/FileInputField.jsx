import PropTypes from 'prop-types';
import React, { useContext } from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { resolveContextOrProp } from '../_helpers/resolveContextOrProp';
import { transferProps } from '../_helpers/transferProps';
import { FormLayoutContext } from '../FormLayout';
import withForwardedRef from '../withForwardedRef';
import styles from './FileInputField.scss';

export const FileInputField = ({
  disabled,
  forwardedRef,
  fullWidth,
  helpText,
  id,
  isLabelVisible,
  label,
  layout,
  required,
  validationState,
  validationText,
  ...restProps
}) => {
  const context = useContext(FormLayoutContext);

  return (
    <label
      className={classNames(
        styles.root,
        fullWidth && styles.isRootFullWidth,
        context && styles.isRootInFormLayout,
        resolveContextOrProp(context && context.layout, layout) === 'horizontal'
          ? styles.rootLayoutHorizontal
          : styles.rootLayoutVertical,
        disabled && styles.isRootDisabled,
        required && styles.isRootRequired,
        getRootValidationStateClassName(validationState, styles),
      )}
      htmlFor={id}
      id={id && `${id}__label`}
    >
      <div
        className={classNames(
          styles.label,
          !isLabelVisible && styles.isLabelHidden,
        )}
        id={id && `${id}__labelText`}
      >
        {label}
      </div>
      <div className={styles.field}>
        <div className={styles.inputContainer}>
          <input
            {...transferProps(restProps)}
            className={styles.input}
            disabled={disabled}
            id={id}
            ref={forwardedRef}
            required={required}
            type="file"
          />
        </div>
        {helpText && (
          <div
            className={styles.helpText}
            id={id && `${id}__helpText`}
          >
            {helpText}
          </div>
        )}
        {validationText && (
          <div
            className={styles.validationText}
            id={id && `${id}__validationText`}
          >
            {validationText}
          </div>
        )}
      </div>
    </label>
  );
};

FileInputField.defaultProps = {
  disabled: false,
  forwardedRef: undefined,
  fullWidth: false,
  helpText: null,
  id: undefined,
  isLabelVisible: true,
  layout: 'vertical',
  required: false,
  validationState: null,
  validationText: null,
};

FileInputField.propTypes = {
  /**
   * If `true`, the input will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Reference forwarded to the `input` element.
   */
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    // eslint-disable-next-line react/forbid-prop-types
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  /**
   * If `true`, the field will span the full width of its parent.
   */
  fullWidth: PropTypes.bool,
  /**
   * Optional help text.
   */
  helpText: PropTypes.node,
  /**
   * ID of the `<input>` HTML element.
   *
   * Also serves as base for ids of nested elements:
   * * `<ID>__label`
   * * `<ID>__labelText`
   * * `<ID>__helpText`
   * * `<ID>__validationText`
   */
  id: PropTypes.string,
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible: PropTypes.bool,
  /**
   * Text field label.
   */
  label: PropTypes.string.isRequired,
  /**
   * Layout of the field.
   *
   * Ignored if the component is rendered within `FormLayout` component
   * as the value is inherited in such case.
   *
   */
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  /**
   * If `true`, the input will be required.
   */
  required: PropTypes.bool,
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  /**
   * Validation message to be displayed.
   */
  validationText: PropTypes.node,
};

export const FileInputFieldWithContext = withForwardedRef(withProviderContext(FileInputField, 'FileInputField'));

export default FileInputFieldWithContext;
