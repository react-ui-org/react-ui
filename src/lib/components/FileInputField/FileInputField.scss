@use "../../styles/tools/form-fields/box-field-elements";
@use "../../styles/tools/form-fields/box-field-layout";
@use "../../styles/tools/form-fields/foundation";
@use "../../styles/tools/form-fields/variants";
@use "../../styles/tools/accessibility";

// Foundation
.root {
    @include foundation.root();
}

.label {
    @include foundation.label();
}

.inputContainer {
    @include box-field-elements.input-container();
}

.input {
    @include accessibility.focus-ring();
}

.helpText,
.validationText {
    @include foundation.help-text();
}

.isRootRequired .label {
    @include foundation.label-required();
}

// States
.isRootStateInvalid {
    @include variants.validation(invalid);
}

.isRootStateValid {
    @include variants.validation(valid);
}

.isRootStateWarning {
    @include variants.validation(warning);
}

// Invisible label
.isLabelHidden {
    @include accessibility.hide-text();
}

// Layouts
.rootLayoutVertical,
.rootLayoutHorizontal {
    @include box-field-layout.vertical();
}

.rootLayoutHorizontal {
    @include box-field-layout.horizontal();
}

.isRootFullWidth {
    @include box-field-layout.full-width();
}

.isRootInFormLayout {
    @include box-field-layout.in-form-layout();
}
