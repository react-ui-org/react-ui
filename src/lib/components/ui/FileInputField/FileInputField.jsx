import PropTypes from 'prop-types';
import React from 'react';
import getRootValidationStateClassName from '../../../helpers/getRootValidationStateClassName';
import { withProviderContext } from '../../../provider';
import transferProps from '../../../utils/transferProps';
import withForwardedRef from '../withForwardedRef';
import styles from './FileInputField.scss';

export const FileInputField = ({
  changeHandler,
  disabled,
  forwardedRef,
  fullWidth,
  helpText,
  id,
  inFormLayout,
  isLabelVisible,
  label,
  layout,
  required,
  validationState,
  validationText,
  ...restProps
}) => (
  <label
    className={[
      styles.root,
      fullWidth ? styles.isRootFullWidth : '',
      inFormLayout ? styles.isRootInFormLayout : '',
      layout === 'horizontal' ? styles.rootLayoutHorizontal : styles.rootLayoutVertical,
      disabled ? styles.isRootDisabled : '',
      required ? styles.isRootRequired : '',
      getRootValidationStateClassName(validationState, styles),
    ].join(' ')}
    htmlFor={id}
    id={id && `${id}__label`}
  >
    <div
      className={[
        styles.label,
        isLabelVisible ? '' : styles.isLabelHidden,
      ].join(' ')}
      id={id && `${id}__labelText`}
    >
      {label}
    </div>
    <div className={styles.field}>
      <div className={styles.inputContainer}>
        <input
          {...transferProps(restProps)}
          className={styles.input}
          disabled={disabled}
          id={id}
          onChange={changeHandler}
          ref={forwardedRef}
          required={required}
          type="file"
        />
      </div>
      {helpText && (
        <div
          className={styles.helpText}
          id={id && `${id}__helpText`}
        >
          {helpText}
        </div>
      )}
      {validationText && (
        <div
          className={styles.validationText}
          id={id && `${id}__validationText`}
        >
          {validationText}
        </div>
      )}
    </div>
  </label>
);

FileInputField.defaultProps = {
  changeHandler: null,
  disabled: false,
  forwardedRef: undefined,
  fullWidth: false,
  helpText: null,
  id: undefined,
  inFormLayout: false,
  isLabelVisible: true,
  layout: 'vertical',
  required: false,
  validationState: null,
  validationText: null,
};

FileInputField.propTypes = {
  /**
   * Function to call when the input has changed.
   */
  changeHandler: PropTypes.func,
  /**
   * If `true`, the input will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Reference forwarded to the `input` element.
   */
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    // eslint-disable-next-line react/forbid-prop-types
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  /**
   * If `true`, the field will span the full width of its parent.
   */
  fullWidth: PropTypes.bool,
  /**
   * Optional help text.
   */
  helpText: PropTypes.node,
  /**
   * ID of the input HTML element. It also serves as a prefix for important inner elements:
   * `<ID>__label`, `<ID>__labelText`, `<ID>__helpText`, and `<ID>__validationText`.
   */
  id: PropTypes.string,
  /**
   * Treat the field differently when it's inside a FormLayout. Do not set manually!
   */
  inFormLayout: PropTypes.bool,
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible: PropTypes.bool,
  /**
   * Text field label.
   */
  label: PropTypes.string.isRequired,
  /**
   * Layout of the field.
   */
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  /**
   * If `true`, the input will be required.
   */
  required: PropTypes.bool,
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  /**
   * Validation message to be displayed.
   */
  validationText: PropTypes.node,
};

export const FileInputFieldWithContext = withForwardedRef(withProviderContext(FileInputField, 'FileInputField'));

export default FileInputFieldWithContext;
