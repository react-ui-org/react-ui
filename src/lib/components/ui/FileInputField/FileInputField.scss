@use '../../../styles/tools/form-fields/box-field-elements';
@use '../../../styles/tools/form-fields/box-field-layout';
@use '../../../styles/tools/form-fields/foundation';
@use '../../../styles/tools/form-fields/relationships';
@use '../../../styles/tools/form-fields/variants';
@use '../../../styles/tools/accessibility';

// Foundation
.root {
  @include foundation.root();
  @include relationships.horizontal-neighbor();
}

.label {
  @include foundation.label();
}

.inputContainer {
  @include box-field-elements.input-container();
}

.input {
  @include accessibility.focus-ring();
}

.helpText,
.validationText {
  @include foundation.help-text();
}

.isRootRequired .label {
  @include foundation.label-required();
}

// States
.isRootStateInvalid {
  @include variants.validation(invalid);
}

.isRootStateValid {
  @include variants.validation(valid);
}

.isRootStateWarning {
  @include variants.validation(warning);
}

// Invisible label
.isLabelHidden {
  @include accessibility.hide-text();
}

// Layouts
.rootLayoutVertical,
.rootLayoutHorizontal {
  @include box-field-layout.vertical();
}

.rootLayoutHorizontal {
  @include box-field-layout.horizontal();
}

.isRootFullWidth {
  @include box-field-layout.full-width();
  @include relationships.vertical-neighbor();
}

.isRootInFormLayout {
  @include box-field-layout.in-form-layout();
}
