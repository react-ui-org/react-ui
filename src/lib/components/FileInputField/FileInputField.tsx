import React, {
  useContext,
} from 'react';
import { withGlobalProps } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { resolveContextOrProp } from '../_helpers/resolveContextOrProp';
import { transferProps } from '../_helpers/transferProps';
import { FormLayoutContext } from '../FormLayout';
import styles from './FileInputField.scss';
import { FileInputFieldProps } from './FileInputField.types';

export const FileInputField: React.FunctionComponent<FileInputFieldProps> = React.forwardRef<HTMLInputElement,
FileInputFieldProps>(({
  disabled = false,
  fullWidth = false,
  helpText,
  id,
  isLabelVisible = true,
  label,
  layout = 'vertical',
  required = false,
  validationState,
  validationText,
  ...restProps
}, ref) => {
  const context = useContext(FormLayoutContext);

  return (
    <label
      className={classNames(
        styles.root,
        fullWidth && styles.isRootFullWidth,
        context && styles.isRootInFormLayout,
        resolveContextOrProp(context && context.layout, layout) === 'horizontal'
          ? styles.isRootLayoutHorizontal
          : styles.isRootLayoutVertical,
        disabled && styles.isRootDisabled,
        required && styles.isRootRequired,
        getRootValidationStateClassName(validationState, styles),
      )}
      htmlFor={id}
      id={id && `${id}__label`}
    >
      <div
        className={classNames(
          styles.label,
          !isLabelVisible && styles.isLabelHidden,
        )}
        id={id && `${id}__labelText`}
      >
        {label}
      </div>
      <div className={styles.field}>
        <div className={styles.inputContainer}>
          <input
            {...transferProps(restProps)}
            disabled={disabled}
            id={id}
            ref={ref}
            required={required}
            type="file"
          />
        </div>
        {helpText && (
          <div
            className={styles.helpText}
            id={id && `${id}__helpText`}
          >
            {helpText}
          </div>
        )}
        {validationText && (
          <div
            className={styles.validationText}
            id={id && `${id}__validationText`}
          >
            {validationText}
          </div>
        )}
      </div>
    </label>
  );
});

export const FileInputFieldWithGlobalProps = withGlobalProps(FileInputField, 'FileInputField');

export default FileInputFieldWithGlobalProps;
