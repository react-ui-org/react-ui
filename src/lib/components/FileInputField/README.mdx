---
name: FileInputField
menu: 'Inputs'
route: /components/file-input-field
---

# FileInputField

FileInputField allows users to select files to upload.

import {
  Playground,
  Props,
} from 'docz'
import { FileInputField } from './FileInputField'

## Basic Usage

To implement the FileInputField component, you need to import it first:

```js
import { FileInputField } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <FileInputField label="Attachment" />
</Playground>

See [API](#api) for all available options.

## General Guidelines

- Apply [native HTML attributes](#forwarding-html-attributes) to control **what
  and how many files** users can select.

- Use **short and descriptive labels**, ideally nouns rather than seemingly
  polite phrases like _Please attach your file here_. Short labels will help
  your users accomplish their task faster.

- Only make the FileInputField's label invisible when there is **another visual
  clue** to guide users through filling the input.

- When a short label is not enough, use **help texts to guide users** before
  they enter anything.

- Use **clear, calm error messages** when there's a problem with what they
  entered. Be positive and focus on solutions to make the error message helpful.

ðŸ‘‰ FileInputField implements native HTML `input type="file"`. Therefore, it is
not possible to affect neither its translations (they depend on a browser), nor
appearance.
[Custom translations and appearance](https://github.com/react-ui-org/react-ui/issues/244)
are considered to be implemented at some point in the future. However, from the
layout perspective, FileInputFields work just like any other form fields.

## Sizes

Full-width fields span the full width of a parent:

<Playground>
  <FileInputField
    fullWidth
    label="First name"
  />
</Playground>

## Forwarding HTML Attributes

When you want to refine your file input fields even further, you
can **add whatever HTML attribute you like.** All attributes that don't
interfere with [component's API](#api) are forwarded to the native HTML input.

<Playground>
  <FileInputField
    accept=".pdf,.jpg,.jpeg,.png"
    helpText={`
      Choose up to 10 files. Allowed extensions are .pdf, .jpg, .jpeg, or .png.
      Size limit is 10 MB.
    `}
    label="Attachment"
    multiple
  />
</Playground>

ðŸ‘‰ Refer to the MDN reference for the full list of supported attributes of the
[file input](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file)
type.

## Invisible Label

In some cases, it may be convenient to visually hide the field label. The label
remains accessible to assistive technologies.

While it may be acceptable for simple forms with just a few fields, it's
dangerous to hide labels from users in most cases. Keep in mind you should
**provide another visual clue** so users know what to fill into the input.

<Playground>
  <FileInputField
    isLabelVisible={false}
    label="Attachment"
  />
</Playground>

## Horizontal Layout

The default vertical layout is very easy to use and work with. However, there
are situations where horizontal layout suits better â€” and that's why React UI
supports this kind of layout as well.

<Playground>
  <FileInputField
    label="Attachment"
    layout="horizontal"
  />
  <FileInputField
    fullWidth
    isLabelVisible={false}
    label="Attachment"
    layout="horizontal"
  />
</Playground>

## Help Text

You may provide an additional help text to clarify how the input should be
filled.

<Playground>
  <FileInputField
    helpText="Choose one or more files to upload."
    label="Attachment"
  />
  <FileInputField
    helpText="Choose one or more files to upload."
    label="Attachment"
    layout="horizontal"
  />
  <FileInputField
    fullWidth
    helpText="Choose one or more files to upload."
    label="Attachment"
    layout="horizontal"
  />
</Playground>

## States

### Validation States

Validation states visually present the result of validation of the input. You
should always **provide a validation message for states other than valid** so
users know what happened and what action they should take or what options they
have.

<Playground>
  <FileInputField
    label="Attachment"
    validationState="valid"
    validationText="Looks good!"
  />
  <FileInputField
    label="Attachment"
    validationState="invalid"
    validationText="Your file is too big. Please select something smaller."
  />
  <FileInputField
    label="Attachment"
    validationState="warning"
    validationText={`
      You selected more than 10 files.
      It may take some time before we process them.
    `}
  />
</Playground>

### Disabled State

It's possible to disable the whole input.

<Playground>
  <FileInputField
    disabled
    label="Attachment"
  />
</Playground>

## API

In addition to the options below, you can specify [React synthetic events] or
any custom HTML attributes that do not interfere with the API, and they will be
passed to the `input` HTML element. This enables making the component
interactive and helps improve its [accessibility](#forwarding-html-attributes).

<Props table of={FileInputField} />

## Theming

Head to [Forms Theming](/customize/theming/forms) to see shared form theming
options.

[React synthetic events]: https://reactjs.org/docs/events.html
