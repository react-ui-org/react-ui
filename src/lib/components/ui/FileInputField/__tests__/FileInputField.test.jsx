import React from 'react';
import {
  shallow,
  mount,
} from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import { FileInputField } from '../FileInputField';

describe('rendering', () => {
  it('renders correctly mandatory props only', () => {
    const tree = shallow(<FileInputField label="label" />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with hidden label', () => {
    const tree = shallow(<FileInputField
      isLabelVisible={false}
      label="With hidden label"
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with custom props', () => {
    const tree = shallow(<FileInputField
      accept=".pdf,.jpg,.jpeg,.png"
      label="With custom props"
      multiple
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(<FileInputField
      disabled
      fullWidth
      helpText="some help"
      id="test"
      inFormLayout
      isLabelVisible={false}
      label="All props"
      layout="horizontal"
      required
      validationState="invalid"
      validationText="some error"
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls changeHandler()', () => {
    const spy = sinon.spy();
    const component = mount(<FileInputField
      changeHandler={spy}
      label="label"
    />);

    component
      .find('input')
      .simulate('change', {
        preventDefault: () => {},
        target: { files: ['test.jpg'] },
      });
    expect(spy.calledOnce).toEqual(true);
  });
});
