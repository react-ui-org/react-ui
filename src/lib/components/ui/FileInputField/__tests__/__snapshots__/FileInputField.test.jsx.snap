// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<label
  className="root   rootLayoutVertical   "
  htmlFor="test"
  id="test__label"
>
  <div
    className="label "
    id="test__labelText"
  >
    label
  </div>
  <div
    className="field"
  >
    <div
      className="inputContainer"
    >
      <input
        className="input"
        disabled={false}
        id="test"
        onChange={null}
        required={false}
        type="file"
      />
    </div>
  </div>
</label>
`;

exports[`rendering renders correctly with all props 1`] = `
<label
  className="root isRootFullWidth isRootInFormLayout rootLayoutHorizontal isRootDisabled isRootRequired isRootStateInvalid"
  htmlFor="test"
  id="test__label"
>
  <div
    className="label isLabelHidden"
    id="test__labelText"
  >
    All props
  </div>
  <div
    className="field"
  >
    <div
      className="inputContainer"
    >
      <input
        className="input"
        disabled={true}
        id="test"
        onChange={null}
        required={true}
        type="file"
      />
    </div>
    <div
      className="helpText"
      id="test__helpText"
    >
      some help
    </div>
    <div
      className="validationText"
      id="test__validationText"
    >
      some error
    </div>
  </div>
</label>
`;

exports[`rendering renders correctly with custom props 1`] = `
<label
  className="root   rootLayoutVertical   "
  htmlFor="test"
  id="test__label"
>
  <div
    className="label "
    id="test__labelText"
  >
    With custom props
  </div>
  <div
    className="field"
  >
    <div
      className="inputContainer"
    >
      <input
        accept=".pdf,.jpg,.jpeg,.png"
        className="input"
        disabled={false}
        id="test"
        multiple={true}
        onChange={null}
        required={false}
        type="file"
      />
    </div>
  </div>
</label>
`;

exports[`rendering renders correctly with hidden label 1`] = `
<label
  className="root   rootLayoutVertical   "
  htmlFor="test"
  id="test__label"
>
  <div
    className="label isLabelHidden"
    id="test__labelText"
  >
    With hidden label
  </div>
  <div
    className="field"
  >
    <div
      className="inputContainer"
    >
      <input
        className="input"
        disabled={false}
        id="test"
        onChange={null}
        required={false}
        type="file"
      />
    </div>
  </div>
</label>
`;
